cdc/owner/feed_state_manager.go

```diff
@@ -484,96 +519,46 @@ func (m *feedStateManager) handleError(errs ...*model.RunningError) {
-		if err.IsChangefeedUnRetryableError() {
-			m.state.PatchInfo(func(info *model.ChangeFeedInfo) (*model.ChangeFeedInfo, bool, error) {
-				if info == nil {
-					return nil, false, nil
-				}
-				info.Error = err
-				return info, true, nil
-			})
-			m.shouldBeRunning = false
-			m.patchState(model.StateError)
-			return
-		}
-	}
-
+	lastError := errs[len(errs)-1]
+	// patch the last error to changefeed info
 	m.state.PatchInfo(func(info *model.ChangeFeedInfo) (*model.ChangeFeedInfo, bool, error) {
 		if info == nil {
 			return nil, false, nil
 		}
-		for _, err := range errs {
-			info.Error = err
-		}
-		return info, len(errs) > 0, nil
+		info.Error = lastError
+		return info, lastError != nil, nil
 	})
 
-	// If we enter into an abnormal state ('error', 'failed') for this changefeed now
+	// if any error is occurred in this tick, we should set the changefeed state to warning
+	// and stop the changefeed
+	if lastError != nil {
+		log.Warn("changefeed meets an error", zap.Any("error", lastError))
+		m.shouldBeRunning = false
+		m.patchState(model.StatePending)
+	}
+
+	// If we enter into an abnormal state 'pending' for this changefeed now
 	// but haven't seen abnormal states in a sliding window (512 ticks),
 	// it can be assumed that this changefeed meets a sudden change from a stable condition.
 	// So we can reset the exponential backoff and re-backoff from the InitialInterval.
 	// TODO: this detection policy should be added into unit test.
-	if len(errs) > 0 {
-		m.lastErrorTime = time.Now()
-		if m.isChangefeedStable() {
-			m.resetErrBackoff()
-		}
-	} else {
-		if m.state.Info.State == model.StateNormal {
-			m.lastErrorTime = time.Unix(0, 0)
-		}
-	}
-	m.shiftStateWindow(m.state.Info.State)
-
-	if m.lastErrorTime == time.Unix(0, 0) {
-		return
-	}
-
-	if time.Since(m.lastErrorTime) < m.backoffInterval {
-		m.shouldBeRunning = false
-		m.patchState(model.StateError)
-	} else {
-		oldBackoffInterval := m.backoffInterval
-
-		m.backoffInterval = m.errBackoff.NextBackOff()
-		m.lastErrorTime = time.Unix(0, 0)
-
-		// NextBackOff() will return -1 once the MaxElapsedTime has elapsed.
-		if m.backoffInterval == m.errBackoff.Stop {
-			log.Warn("The changefeed won't be restarted "+
-				"as it has been experiencing failures for "+
-				"an extended duration",
-				zap.Duration(
-					"maxElapsedTime",
-					m.errBackoff.MaxElapsedTime,
-				),
-			)
-			m.shouldBeRunning = false
-			m.patchState(model.StateFailed)
-			return
-		}
-
-		log.Info("changefeed restart backoff interval is changed",
-			zap.String("namespace", m.state.ID.Namespace),
-			zap.String("changefeed", m.state.ID.ID),
-			zap.Duration("oldInterval", oldBackoffInterval),
-			zap.Duration("newInterval", m.backoffInterval))
+	if m.isChangefeedStable() {
+		m.resetErrRetry()
 	}
 }
 
 func (m *feedStateManager) handleWarning(errs ...*model.RunningError) {
+	if len(errs) == 0 {
+		return
+	}
+	lastError := errs[len(errs)-1]
+	m.patchState(model.StateWarning)
```

Comment:
Perhaps lastRetryCheckpointTs should be updated before pathch warning state?

Reply:
Changefeed will not retry on a warning, warning will be retried on sink level.

Reply:
Changefeed will not retry on a warning, warning will be retried on sink level.